packages/@aws-cdk/aws-ecs/lib/container-image.ts

```diff
@@ -2,29 +2,11 @@ import ecr = require('@aws-cdk/aws-ecr');
 import cdk = require('@aws-cdk/cdk');
 
 import { ContainerDefinition } from './container-definition';
-import { AssetImage, AssetImageProps } from './images/asset-image';
-import { DockerHubImage } from './images/dockerhub';
-import { EcrImage } from './images/ecr';
-
-/**
- * A container image
- */
-export interface IContainerImage {
-  /**
-   * Name of the image
-   */
-  readonly imageName: string;
-
-  /**
-   * Called when the image is used by a ContainerDefinition
-   */
-  bind(containerDefinition: ContainerDefinition): void;
-}
 
 /**
  * Constructs for types of container images
  */
-export class ContainerImage {
+export abstract class ContainerImage {
```

Comment:
I'm not sure I agree with this change, nor that it is strictly necessary for deCDK. Surely you could implement some heuristics around this.

Reply:
I have an idea how we can do this (convert "add"s to prop arrays) but that's not implemented. Also, this approach doesn't rely on the fact that users call `add` to associate a container to the task definition, they can either use `add` or "new" the ContainerDefinition themselves, and it will "register" itself on the task definition.

Reply:
Regardless of deCDK, I think this is a better API because it's easier to discover the various options from the type of the options. It is also more aligned with other APIs like `lambda.Code` and `lambda.Runtime`. Any reason to insist on using an interface instead of an abstract class?

Reply:
Interfaces give us more flexibility to make implementation choices, classes lock us into a single-inheritance model that at some point is not going to fit whatever we want to do.

I know this is wishy-washy, I don't have a hard example to convince you right now. So fine, I guess.

I also question the necessity of making this change right this instant. When are we productizing deCDK?

Reply:
The reason I need this is in order for the deCDKs prototype which I would like to release to include an ECS example.

I think standardizing on how we implement these “unions” pattern will be useful because it’s reoccurring. I understand the benefit of using interfaces and maybe that should be the pattern but I also find the discoverability when using these a bit problematic. Using a class with a bunch of static methods/properties is closer to enums in discoverability so that’s where the value is in my mind. 

Perhaps we can enforce via awslint that the jsdoc will include some information about possible implementations?